core/src/main/java/org/frankframework/pipes/ReplacingInputStream.java

```diff
@@ -0,0 +1,171 @@
+ * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+ * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+public class ReplacingInputStream extends FilterInputStream {
+
+	private final boolean allowUnicodeSupplementaryCharacters;
+	private final Queue<Integer> inQueue;
+	private final Queue<Integer> outQueue;
+	private final String replaceNonXmlChar;
+	private final boolean replaceNonXmlChars;
+	private final byte[] replacement;
+	private final byte[] search;
+
+	public ReplacingInputStream(InputStream in, String search, String replacement, boolean replaceNonXmlChars,
+								String replaceNonXmlChar, boolean allowUnicodeSupplementaryCharacters) {
+
+		super(in);
+		this.replaceNonXmlChars = replaceNonXmlChars;
+		this.replaceNonXmlChar = replaceNonXmlChar == null ? "" : replaceNonXmlChar;
+		this.allowUnicodeSupplementaryCharacters = allowUnicodeSupplementaryCharacters;
+
+		this.inQueue = new LinkedList<>();
+		this.outQueue = new LinkedList<>();
+
+		this.search = (search == null) ? "".getBytes() : search.getBytes();
+		this.replacement = (replacement == null) ? "".getBytes() : replacement.getBytes();
+	}
+
+	@Override
+	public int read() throws IOException {
+		while (outQueue.isEmpty()) {
+			readAhead();
+
+			int length = search.length;
+			if (length > 0 && isMatchFound()) {
+				IntStream.range(0, length)
+						.forEach(a -> inQueue.remove());
+
+				IntStream.range(0, replacement.length)
+						.forEach(b -> outQueue.offer((int) replacement[b]));
+			} else {
+				outQueue.add(inQueue.remove());
+			}
+		}
+
+		return outQueue.remove();
+	}
+
+	@Override
+	public int read(byte[] b) throws IOException {
+		return read(b, 0, b.length);
+	}
+
+	/**
+	 * copied straight from InputStream implementation, just needed to use {@link #read()} from this class
+	 * @see InputStream#read(byte[], int, int)
+	 */
+	@Override
+	public int read(@Nonnull byte[] b, int off, int len) throws IOException {
+		if (off < 0 || len < 0 || len > b.length - off) {
+			throw new IndexOutOfBoundsException();
+		} else if (len == 0) {
+			return 0;
+		}
+
+		int c = read();
+		if (c == -1) {
+			return -1;
+		}
+		b[off] = (byte) c;
+
+		int i = 1;
+		try {
+			for (; i < len; i++) {
+				c = read();
+				if (c == -1) {
+					break;
+				}
+				b[off + i] = (byte) c;
+			}
+		} catch (IOException ee) {
+		}
+		return i;
+	}
+
+	private int getNextValue() throws IOException {
+		int next = super.read();
+
+		if (next != -1 && replaceNonXmlChars && !XmlEncodingUtils.isPrintableUnicodeChar(next, allowUnicodeSupplementaryCharacters)) {
+			if (!StringUtils.isEmpty(replaceNonXmlChar)) {
+				next = replaceNonXmlChar.charAt(0);
```

Comment:
Aangezien de replacement nooit meer dan 1 char kan zijn, kunnen we die intern misschien ook als `int` opslaan ipv `String`. Dan hoeven we niet iedere keer `.charAt(0)` te doen, wat weer overhead geeft.